django_cloud_tasks/models.py

```diff
@@ -0,0 +1,133 @@
+        for routine in routines:
+            routine.revert()
+
+    def create_routines_and_vertex(self, routines: List[Dict]) -> QuerySet:
+        previous_routine = None
+        for routine_payload in routines:
+            routine = self.routines.create(**routine_payload)
+            if previous_routine:
+                RoutineVertex.objects.create(routine=previous_routine, next_routine=routine)
+            previous_routine = routine
+
+        return self.routines
+
+
+class Routine(models.Model):
+    class Statuses(models.TextChoices):
+        PENDING = ("pending", "Pending")
+        SCHEDULED = ("scheduled", "Scheduled")
+        RUNNING = ("running", "Running")
+        COMPLETED = ("completed", "Completed")
+        FAILED = ("failed", "Failed")
+        REVERTING = ("reverting", "Reverting")
+        REVERTED = ("reverted", "Reverted")
+        ABORTED = ("aborted", "Aborted")
+
+    task_name = models.CharField(max_length=100)
+    pipeline = models.ForeignKey(
+        to="django_cloud_tasks.Pipeline",
+        related_name="routines",
+        on_delete=models.PROTECT,
+    )
+    body = models.JSONField(
+        default=dict,
+        encoder=serializers.JSONEncoder,
+    )
+    attempt_count = models.PositiveIntegerField(default=0)
+    max_retries = models.PositiveIntegerField(null=True)
+    output = models.JSONField(
+        null=True,
+        blank=True,
+        encoder=serializers.JSONEncoder,
+    )
+    starts_at = models.DateTimeField(null=True, blank=True)
+    ends_at = models.DateTimeField(null=True, blank=True)
+    status = models.CharField(
+        max_length=20,
+        choices=Statuses.choices,
+        default=Statuses.PENDING,
+    )
+    created_at = models.DateTimeField(
+        auto_now_add=True,
+    )
+    updated_at = models.DateTimeField(
+        auto_now=True,
+    )
+    next_routines = models.ManyToManyField(
+        to="Routine",
+        through="RoutineVertex",
+        through_fields=("routine", "next_routine"),
+        related_name="dependent_routines",
+    )
+
+    def fail(self, output: str):
+        self.output = output
+        self.status = self.Statuses.FAILED
+        self.ends_at = datetime.now()
+        self.save()
+
+    def complete(self, output: str):
+        self.output = output
+        self.status = self.Statuses.COMPLETED
+        self.ends_at = datetime.now()
+        self.save()
+
+    def enqueue(self):
+        with transaction.atomic():
+            self.status = self.Statuses.SCHEDULED
+            self.starts_at = datetime.now()
+            self.save()
+
+    def revert(self):
+        with transaction.atomic():
+            if self.status == self.Statuses.COMPLETED:
+                self.status = self.Statuses.REVERTING
+                self.save()
+                return
+
+            self.status = self.Statuses.ABORTED
+            self.save()
```

Comment:
What about RUNNING and PENDING statuses?

Reply:
RUNNING IDK how to handle right now
Pending will be aborted.

Reply:
Same comment about the `TODO: `